### PR TITLE
8336267: Method and Constructor signature parsing can be shared on the root object

### PR DESCRIPTION
**Pull request by liach**
A straightforward optimization, to share the signature parsing of method, constructor, and field between the root and the copied objects, like how method handle accessors are shared.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336267](https://bugs.openjdk.org/browse/JDK-8336267): Method and Constructor signature parsing can be shared on the root object (**Enhancement** - P4)

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20179/head:pull/20179` \
`$ git checkout pull/20179`

Update a local copy of the PR: \
`$ git checkout pull/20179` \
`$ git pull https://git.openjdk.org/jdk.git pull/20179/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20179`

View PR using the GUI difftool: \
`$ git pr show -t 20179`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20179.diff">https://git.openjdk.org/jdk/pull/20179.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20179#issuecomment-2227669452)
</details>

### CODE REVIEW

**Comment by liach** [comment]
Keep alive.

**Comment by cl4es** [comment]
There are a lot of transient volatile fields in this code. Could anything be done to make the various 'Repository` classes more `final` and less `transient`/`volatile` here so that they can be safely published without synchronization, making it more obviously correct that we can safely pick them up from a shared root node (which might more typically be accessed concurrently and thus more exposed to races)?

**Comment by liach** [comment]
Yep, there's #19281 exactly for that purpose: https://github.com/openjdk/jdk/pull/19281/files#diff-5106bd8a6cf3197be7eaee467983c8992981a615bedb66efa50dd97c60cb0041 Uses `volatile @Stable` instead (or just `StableValue` once #19625 is ready)

File: src/java.base/share/classes/java/lang/reflect/Constructor.java [hunk 1]
```diff
@@ -90,10 +90,15 @@ ConstructorRepository getGenericInfo() {
...
-                ConstructorRepository.make(getSignature(),
-                                           getFactory());
+            var root = this.root;
+            if (root != null) {
+                genericInfo = root.getGenericInfo();
+            } else {
+                // create and cache generic info repository
+                genericInfo =
+                        ConstructorRepository.make(getSignature(),
+                                getFactory());
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/reflect/Constructor.java:100]:
```suggestion
                genericInfo = ConstructorRepository.make(getSignature(), getFactory());
```

Nit: This fits well in one line.

File: src/java.base/share/classes/java/lang/reflect/Field.java [hunk 2]
```diff
@@ -109,9 +109,14 @@ private FieldRepository getGenericInfo() {
...
-            // create and cache generic info repository
-            genericInfo = FieldRepository.make(getGenericSignature(),
-                                               getFactory());
+            var root = this.root;
+            if (root != null) {
+                genericInfo = root.getGenericInfo();
+            } else {
+                // create and cache generic info repository
+                genericInfo = FieldRepository.make(getGenericSignature(),
+                        getFactory());
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/reflect/Field.java:118]:
```suggestion
                genericInfo = FieldRepository.make(getGenericSignature(), getFactory());
```

File: src/java.base/share/classes/java/lang/reflect/Method.java [hunk 3]
```diff
@@ -113,9 +113,15 @@ MethodRepository getGenericInfo() {
...
-            // create and cache generic info repository
-            genericInfo = MethodRepository.make(getGenericSignature(),
-                                                getFactory());
+            var root = this.root;
+            if (root != null) {
+                genericInfo = root.getGenericInfo();
+            } else {
+                // create and cache generic info repository
+                genericInfo = MethodRepository.make(getGenericSignature(),
+                        getFactory());
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/reflect/Method.java:122]:
```suggestion
                genericInfo = MethodRepository.make(getGenericSignature(), getFactory());
```

**Review by mlchung — COMMENTED** [review]
It's okay to extend the root object to share `genericInfo` besides the accessor.  But we need to update the comment to make that clear.

```
    @Stable
    private ConstructorAccessor constructorAccessor;
    // For sharing of ConstructorAccessors. This branching structure
    // is currently only two levels deep (i.e., one root Constructor
    // and potentially many Constructor objects pointing to it.)
    //
    // If this branching structure would ever contain cycles, deadlocks can
    // occur in annotation code.
    private Constructor<T>      root;
```

The comment for `root` should be extended to cover the generics repository.   I also suggest to move the declaratoin of instance variable `genericInfo` closer to the accessor and `root`.

**Comment by liach** [comment]
Thanks; I have simplified the existing comments and grouped the lazy shareable fields together.

**Review by mlchung — APPROVED** [review]
Looks good.

File: src/java.base/share/classes/java/lang/reflect/Method.java [hunk 4]
```diff
@@ -81,22 +81,20 @@ public final class Method extends Executable {
...
+    /**
+     * Methods are mutable due to {@link AccessibleObject#setAccessible(boolean)}.
+     * Thus, we return a new copy of a root each time a method is returned.
+     * Some lazily initialized immutable states can be stored on root and shared to the copies.
+     */
+    private Method root;
+    private transient volatile MethodRepository genericInfo;
+    private @Stable MethodAccessor methodAccessor;
+    private int hash; // not shared right now, eligible if expensive
+    // End shared states
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/reflect/Method.java:97]:
```suggestion
    // End shared states
    private int hash; // not shared right now, eligible if expensive
```

**Comment by liach** [comment]
Applied this minor patch. java/lang/reflect tests are still all fine.

**Review by mlchung — APPROVED** [review]

**Pull request closed without merging**